### ISSUE TITLE
Bufferbloat example does not work on Ubuntu 18.04 and above

### ISSUE THREAD

**Issue by cw150** [mininet/mininet#1045]
### Expected/Desired Behavior:
Bufferbloat lab exercise (https://github.com/mininet/mininet/wiki/Bufferbloat) should work with the current mininet 2.3.0 VM images based on Ubuntu 18.04 and 20.04.

### Actual Behavior:
Tracing of TCP congestion window cwnd fails, because kernel module tcp_probe is not present on Ubuntu 20.04 (the module is present on 18.04, but cannot be loaded due to missing kernel functions). 
Additionally, on 20.04 it is necessary to first modify the python scripts to work with python 3 (falling back to python 2 is possible, but requires installing matplotlib for python 2 with pip, because it is no longer available as a package.)

### Detailed Steps to Reproduce the Behavior:
Launch `monitor.sh` as per Part 4 of the example. This executes `exp_monitor.py`, which will fail to launch TCP tracing with `modprobe tcp_probe`. The scripts seem to run, but the trace file and the figures generated with `plot_tcpprobe.py` will be empty.

### Additional Information:
The Bufferbloat example works on the current mininet 2.3.0 VM image based on Ubuntu 16.04 (tested with mininet-2.3.0-210211-ubuntu-16.04.7-server-amd64-ovf.zip).
Since tcp_probe has been removed from the kernel, several sources suggest using ftrace instead. This implies modifying plot_tcpprobe.py` to match the new trace file format. However, this fails as well, since mininet network events are not logged by the ftrace mechanism, even after mounting `/sys/kern/debug`. This changed tracing behavior has been observed by others: [https://stackoverflow.com/questions/65868476/how-to-use-ftrace-for-tcp-probe](url). 

I suggest adding a remark to the Bufferbloat exercise, stating that in order to work "out of the box," it requires python 2 and a kernel with working tcp_probe module. The simplest way to achieve this is to use the provided Ubuntu 16.04 VM images (without upgrading the kernel).

**Comment by cw150** [comment]
Hello, our "workaround" was to fall back to Ubuntu 16.04 (we used a 
mininet VM for a networking class, time was short).
We did briefly try what you did and ended up seeing the same behavior. 
It has likely to do with the way mininet virtualizes the network layer 
and/or how ftrace treats physical vs. virtual interfaces. Since tracing 
mechanisms have completely changed in newer kernels (as in Ubuntu 20+), 
I don't think it is worth trying to get it working for just 18.04.

**Comment by lantz** [comment]
I actually updated monitoring to work with ftrace and eBPF, but I haven't had lots of free time to spend on updating the bufferbloat exercise, etc. - that's kind of what we'd rely on the community for.

The lack of a stable kernel API makes monitoring extremely fragile unfortunately. I don't think there is a good long-term solution other than fixing the kernel version.

**Comment by jamestiotio** [comment]
Hi @lantz, apologies for the late reply. I can attempt to implement some kind of fix for Ubuntu 20.04+ in my free time, but I cannot guarantee that it can be fixed properly since I have to look into this more. The change in the kernel API might also even prevent any fixes to be implemented at all since `ftrace` might have an entirely different behaviour internally as compared to `tcp_probe` (especially on how it treats physical and virtual interfaces differently).

To assist with my analysis, can you point me on how you managed to find these data by using `ftrace` and eBPF (here, I assume that the input socket is `sk`, which is of the `sock` struct defined in the kernel):
- `inet_sock`: Calling `inet_sk` with `sk` as the parameter
  - `inet_saddr`
  - `inet_sport`
  - `inet_daddr`
  - `inet_dport`
- Length of `sk_buff` (by calling `tcp_write_queue_tail` with `sk` as the parameter)
- `tcp_sock`: Calling `tcp_sk` with `sk` as the parameter
  - `snd_nxt`
  - `snd_una`
  - `snd_cwnd`
  - `snd_wnd`
  - `rcv_wnd`
  - `srtt_us`
- The value of the current ssthresh (by calling `tcp_current_ssthresh` with `sk` as the parameter)

I believe that we need all of the above data to implement some kind of "imitation" of `tcp_probe` such that the bufferbloat example would work properly.

My current not-working implementation is as such:

```c
#include <linux/module.h>
#include <linux/vmalloc.h>
#include <net/tcp.h>

void tcp_vreno_in_ack_event(struct sock *sk, u32 flags)
{
	const struct tcp_sock *tp = tcp_sk(sk);
	const struct inet_sock *isock = inet_sk(sk);

	uint16_t sport = ntohs(isock->inet_sport);
	uint16_t dport = ntohs(isock->inet_dport);

	struct sk_buff *skb = tcp_write_queue_tail(sk);
	uint16_t length = skb == NULL ? 0 : skb->len;

	trace_printk(KERN_INFO "%pI4:%u %pI4:%u %d %#x %#x %u %u %u %u %u\n",
			&isock->inet_saddr, sport, &isock->inet_daddr, dport, length, tp->snd_nxt, tp->snd_una, tp->snd_cwnd, tcp_current_ssthresh(sk), tp->snd_wnd, tp->rcv_wnd, tp->srtt_us >> 3);
}


struct tcp_congestion_ops tcp_reno_verbose = {
	.flags		= TCP_CONG_NON_RESTRICTED,
	.name		= "reno_verbose",
	.owner		= THIS_MODULE,
	.ssthresh	= tcp_reno_ssthresh,
	.cong_avoid	= tcp_reno_cong_avoid,
	.undo_cwnd	= tcp_reno_undo_cwnd,
	.in_ack_event	= tcp_vreno_in_ack_event,
};

/* Some other statements to register and unregister this custom kernel module */
```

If I am going in the wrong direction with this approach, do also point it out and perhaps I can try to approach this problem from a different angle.

Cheers!

**Comment by janev94** [comment]
@jamestiotio @lantz Hi, I see that my stackoverflow answer has made it here... and so did my verbose reno code (without proper acknowledgments). I do believe that this is the "correct" long-term solution as it would require change in the kernel TCP API in order for it to break. That said, James has only _borrowed_ my New Reno code and not the others for Cubic, etc. I do believe that a series of sysctl handles and better information exposure would make it easier for kernel parsing. Additionally, a makefile and instructions how to load the custom modules is also needed and is lacking from the previous answer.

For future reference, please note that all this code was copyrighted under GPLv2 and I endorse redistribution as long as the correct headers and acknowledgements are given. I would also be happy to provide working kernel modules for whichever algorithms it is that you want to measure bufferbloat with instructions if that is the way you would like to proceed.

**Comment by jamestiotio** [comment]
@janev94 Got it. I have added an explicit reference link to your GitHub Gist in my previous comment.

Just to be clear, the code in my comment was less intended to be a full solution and more for aiding discussion, since said implementation did not work.

I will leave the discussion of fixing the bufferbloat example between you and the maintainers of mininet then.

**Comment by lantz** [comment]
I agree it would be nice to update the buffer bloat example.  This isn't exactly a Mininet issue.
